jobs/nginx/templates/config/nginx.conf

```diff
@@ -68,7 +68,7 @@ http {
   upstream alertmanager {
     <% alertmanager.instances.map do |instance| %>
       <% if p('nginx.alertmanager.cross_zone_load_balancing') || instance.az == spec.az %>
-      server <%= "#{instance.address}:#{alertmanager.p('alertmanager.web.port')} fail_timeout=0;" %>
+      server <%= "#{instance.address}:#{p('nginx.alertmanager.upstream_port', alertmanager.p('alertmanager.web.port'))} fail_timeout=0;" %>
```

Comment:
can you clarify the usage of `nginx.{prometheus,grafana,alertmanager}.upstream_port` ? I don't really get it. 
Even with the possibility to configure the alertmanager's listen address, I don't see a "new port" to send request on from the nginx proxy.

Reply:
This feature is intended to be used along with another job that has the capability to enforce authorization. Let me elaborate on this using the following diagram.

<img src="https://user-images.githubusercontent.com/6535425/143419144-489b617a-1a7a-4295-8352-c35c2ba7ba4f.png" width="496">

As we have an oauth2-proxy deployed together with an alertmanager that is configured to bind to 127.0.0.1:9093, we can restrict access from outside the VM but allow access from oauth2-proxy. In this architecture, the port number that Alertmanager listens to could not even be accessible from nginx too so `nginx.alertmanager.upstream_port` is set to `80` to solve this. Note when oauth2-proxy is deployed together with an nginx job, there is no way to limit access to alertmanager because alertmanager binds to `0.0.0.0`.

Reply:
In this use-case, couldn't the `oauth2-proxy` provide a [bosh link](https://bosh.io/docs/links/) of type `alertmanager`  ? Then gives the link to nginx job to be [consumed](https://github.com/bosh-prometheus/prometheus-boshrelease/blob/43f0d6b7fd82da812769f34e1babe2f3c77788b2/jobs/nginx/spec#L26) ? In my opinion, that is exactly the use-case of bosh links: substitute a natural target by another one that fulfill the same requirements/interface 

There are already so many ports to handle, plus, alertmanager has itself multiple listen ports. My fear is that it will get less and less understandable.

this next question is beyond the scope of the PR: 
- isn't it possible to get the oauth in front (or into) nginx and use mutual-TLS authentication between nginx and `{alertmanager,prometheus,grafana}` ?

Reply:
Ah, thanks for pointing out! This case is completely and naturally implementable by using custom provider definitions in an oauth2-proxy job and I've just confirmed that it works as well with the following setup. Removed them as they just makes the things complicated. 

```yaml
instance_groups:
  - name: alertmanager
    jobs:
      - name: alertmanager
        release: prometheus
        properties:
          web:
            port: 9093
          # ...
       - name: oauth2-proxy
         release: oauth2-proxy
         custom_provider_definitions:
           - name: oauth2-proxy
             type: alertmanager
             properties:
               - alertmanager.web.port
         properties:
           alertmanager:
             web:
               port: 80
           oauth2-proxy:
             http_address: 0.0.0.0:80
             # ...
   - name: nginx
     jobs:
       - name: nginx
         release: prometheus
         consumes:
           alertmanager:
             from: oauth2-proxy
         properties:
           # ...
```

> isn't it possible to get the oauth in front (or into) nginx and use mutual-TLS authentication between nginx and {alertmanager,prometheus,grafana} ?

It could be possible but we need to make `proxy_pass` more control instead.
https://github.com/bosh-prometheus/prometheus-boshrelease/blob/v26.6.0/jobs/nginx/templates/config/alertmanager_location.conf#L2